### PR TITLE
async_base example uses async_initiate

### PR DESCRIPTION
**Pull request by madmongo1**
I have not added any more narrative comments. Do we need to do so now that the op is a separate class and there is an initiator involved?

### CODE REVIEW

File: include/boost/beast/core/async_base.hpp [hunk 1]
```diff
@@ -76,54 +76,73 @@ namespace beast {
 
     @code
 
-    // Asynchronously read into a buffer until the buffer is full, or an error occurs
-    template<class AsyncReadStream, class ReadHandler>
-    typename net::async_result<ReadHandler, void(error_code, std::size_t)>::return_type
-    async_read(AsyncReadStream& stream, net::mutable_buffer buffer, ReadHandler&& handler)
+    template<class DeducedHandler, class AsyncReadStream>
```
Comment (vinniefalco) [inline include/boost/beast/core/async_base.hpp]: A class template parameter is never labeled "deduced" I only use that for function parameters.

File: include/boost/beast/core/async_base.hpp [hunk 2]
```diff
@@ -76,54 +76,73 @@ namespace beast {
...
+    struct async_read_op
+    : async_base<DeducedHandler, typename AsyncReadStream::executor_type>
     {
-        using handler_type = BOOST_ASIO_HANDLER_TYPE(ReadHandler, void(error_code, std::size_t));
-        using base_type = async_base<handler_type, typename AsyncReadStream::executor_type>;
+        AsyncReadStream& stream_;
+        net::mutable_buffer buffer_;
+        std::size_t total_bytes_transferred_;
+
+        async_read_op(
```
Comment (vinniefalco) [inline include/boost/beast/core/async_base.hpp]: Here we need `template<class DeducedHandler>` and `DeducedHandler&&` parameter type

File: include/boost/beast/core/async_base.hpp [hunk 3]
```diff
@@ -76,54 +76,73 @@ namespace beast {
...
-        op(stream, buffer, init.completion_handler);
-        return init.result.get();
+    // Asynchronously read into a buffer until the buffer is full, or an error occurs
+    template<class AsyncReadStream, class ReadHandler>
+    typename net::async_result<ReadHandler, void(error_code, std::size_t)>::return_type
+    async_read(AsyncReadStream& stream, net::mutable_buffer buffer, ReadHandler&& handler)
+    {
+        return
+            net::async_initiate<ReadHandler,
+                                   void(error_code, std::size_t)>(
```
Comment (vinniefalco) [inline include/boost/beast/core/async_base.hpp]: uhh..

File: include/boost/beast/core/async_base.hpp (hunk 2 above)
Comment (madmongo1) [reply include/boost/beast/core/async_base.hpp]: why? pass by value-move into place is a perfectly optimisable idiom.

File: include/boost/beast/core/async_base.hpp [hunk 4]
```diff
@@ -76,54 +76,74 @@ namespace beast {
...
+        std::size_t total_bytes_transferred_;
+
+        template<class DeducedHandler>
+        async_read_op(
+            AsyncReadStream& stream,
+            net::mutable_buffer buffer,
+            DeducedHandler&& handler)
+            : async_base<Handler,
+                typename AsyncReadStream::executor_type>
+                (std::forward<DeducedHandler>(handler), stream.get_executor())
```
Comment (vinniefalco) [inline include/boost/beast/core/async_base.hpp]:
```
            : async_base<Handler,typename AsyncReadStream::executor_type>(
                std::forward<DeducedHandler>(handler), stream.get_executor())
```

File: include/boost/beast/core/async_base.hpp [hunk 5]
```diff
@@ -76,54 +76,74 @@ namespace beast {
...
+    struct async_read_op
+    : async_base<Handler, typename AsyncReadStream::executor_type>
     {
-        using handler_type = BOOST_ASIO_HANDLER_TYPE(ReadHandler, void(error_code, std::size_t));
-        using base_type = async_base<handler_type, typename AsyncReadStream::executor_type>;
+        AsyncReadStream& stream_;
+        net::mutable_buffer buffer_;
+        std::size_t total_bytes_transferred_;
+
+        template<class DeducedHandler>
```
Comment (vinniefalco) [inline include/boost/beast/core/async_base.hpp]: but.. you said this wasn't needed, did you change your mind?

Comment (madmongo1) [reply include/boost/beast/core/async_base.hpp]: I'd started so I felt it easier to finish

**Pull request closed without merging**